### ISSUE TITLE
supporting offset in poisson

### ISSUE THREAD

**Issue by apoorvalal** [py-econometrics/pyfixest#695]
[this](https://en.wikipedia.org/wiki/Poisson_regression#%22Exposure%22_and_offset) gives the gist; having an `offset` argument would involve constraining the coefficient to 1 and being able to model rates (in contrast to the regular use to model counts). Statsmodels supports it under an [`exposure` arg](https://www.statsmodels.org/devel/generated/statsmodels.discrete.discrete_model.Poisson.html#statsmodels.discrete.discrete_model.Poisson).

[more bg](https://stats.stackexchange.com/questions/433674/why-does-adding-an-offset-change-the-coefficients-in-a-poisson-regression)

**Comment by s3alfisc** [comment]
Yes, nice one. Have never used an offset before in my life, so hadn't bothered with this yet 😀 looks like it might not be too hard to add! fixest::feglm() has an "offset" argument, so we should maybe stick with that instead of statsmodels kwarg solution?

**Comment by bradhackinen** [comment]
I'd love to have this feature for a project I am working on. Is there anything I can do to help?

**Comment by s3alfisc** [comment]
Hi @bradhackinen , thanks for bringing this back to my attention! =) Just thought about this for a few minutes, and this is what I think we'd need to do: 

I would add an `offset` argument to the [fepois function](https://github.com/py-econometrics/pyfixest/blob/03693d0ec8dc283ff00de2cdf210e886c521737a/pyfixest/estimation/estimation.py#L502), following the fixest example.

From the `r-fixest` docs: 

> **offset** 
>     A formula or a numeric vector. An offset can be added to the estimation. If equal to a formula, it should be of the form (for example) ~0.5*x**2. This offset is linearly added to the elements of the main formula 'fml'.
 
For the beginning, I would only support a scalar input, no formulas. 

We'd first have to add the `offset` argument to [fepois](https://github.com/py-econometrics/pyfixest/blob/master/pyfixest/estimation/fepois_.py). From there, we pass it via [FixestMulti ](https://github.com/py-econometrics/pyfixest/blob/03693d0ec8dc283ff00de2cdf210e886c521737a/pyfixest/estimation/FixestMulti_.py#L334)to the Fepois method, and then we'd have to adjust for it in every WLS fit in the [iterated weighted least squares fitting ](https://github.com/py-econometrics/pyfixest/blob/03693d0ec8dc283ff00de2cdf210e886c521737a/pyfixest/estimation/fepois_.py#L238) step. 

Would you be up to take a stab at this @bradhackinen? Of course I'm more than happy to help along the way =)

**Comment by bradhackinen** [comment]
Yeah, I was looking at the code yesterday and I think I can figure out most of it. Adding an argument for a scalar input is no problem, but I'm still working on understanding how to modify the WLS fit, particularly how that works with the demeaning step. If you have a clear idea of how that should be done, I'd be glad to hear some hints. Otherwise I can probably figure it out eventually by deciphering the fixest R code and reading Laurent Berge's paper.

**Comment by s3alfisc** [comment]
Alright cool, I'll try to give it some though tomorrow / in the next days and put down some notes!

**Comment by bradhackinen** [comment]
That would be very helpful, thanks! I'll also take another look at it on Monday and see what I can figure out on my own.

**Comment by s3alfisc** [comment]
So either I am completely off or this might actually be kind of straightforward, at least for the case with no fixed effects? 🤔 

We simply adjust `reg_Z = reg_Z - offset` on top of line https://github.com/py-econometrics/pyfixest/blob/03693d0ec8dc283ff00de2cdf210e886c521737a/pyfixest/estimation/fepois_.py#L267 - hence subtract the offset from the dependent variable in linear space. 

Then, everything stays as is for a while until we compute `eta`, where we have to add back the offset: 

```python
eta = Z - resid + offset
```
Maybe we can build a working solution without fixed effect demeaning first and then iterate on that? And maybe my brain is clouded right now, though I think the above would work with fixed effect demeaning too?

**Comment by s3alfisc** [comment]
On the point above: I'm more convinced that it should work with fixed effects too, as a constant offset will be absorbed by the FE. Basically fixed effects Poisson is invariant to constant offsets. Here is a `fixest` example: 

```r
library(fixest)
data(mtcars)
head(mtcars)

fit1 = fepois(cyl ~ mpg | hp, offset = 10, data = mtcars)
fit2 = fepois(cyl ~ mpg | hp, offset = 0, data = mtcars)
fit3 = fepois(cyl ~ mpg | hp, data = mtcars)

etable(fit1, fit2, fit3, digits = 6)
                                fit1                 fit2                 fit3
Dependent Var.:                  cyl                  cyl                  cyl
                                                                              
mpg             -0.011859 (0.014323) -0.011859 (0.014323) -0.011859 (0.014323)
Fixed-Effects:  -------------------- -------------------- --------------------
hp                               Yes                  Yes                  Yes
_______________ ____________________ ____________________ ____________________
S.E.: Clustered               by: hp               by: hp               by: hp
Observations                      32                   32                   32
Squared Cor.                 0.97415              0.97415              0.97415
Pseudo R2                    0.12166              0.12166              0.12166
BIC                           196.74               196.74               196.74
---
Signif. codes: 0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

```

**Comment by bradhackinen** [comment]
Useful hints, thanks. That makes sense to me for the no-FE case. Intuitively, I would think you need to demean the offset in the FE case. Or maybe transforming Z works there too? I don't have a clear understanding of how the demeaning works in the non-linear model yet.

**Comment by bradhackinen** [comment]
I think I have it working. The simple modification @s3alfisc proposed seems to work in both the FE and non-FE case. Here is a [pull request](https://github.com/py-econometrics/pyfixest/pull/805) that implements the feature.

I tested it manually on a few different cases, and the output looks correct in both the FE and non-FE cases. I don't have r2py set up, so I didn't try to update `test_poisson.py`, but I can share my (very simple) test scripts if that would be helpful.

**Comment by s3alfisc** [comment]
Hi @bradhackinen, awesome, thanks you! I'll take a look this evening and will add a few tests. Thank you!

**Comment by s3alfisc** [comment]
Finally merged with https://github.com/py-econometrics/pyfixest/commit/50e58c9c09ea0d1abe7b458d9772c12bddc18e6f.

**Comment by bradhackinen** [comment]
Thanks @s3alfisc! I'm looking forward to replacing my local version of pyfixest with the official distribution in my project replication code.

**Comment by s3alfisc** [comment]
> Thanks @s3alfisc! I'm looking forward to replacing my local version of pyfixest with the official distribution in my project replication code.

We will try our best for a quick release, but it might take a few more days - the main branch currently contains a breaking API change and we have decided to adjust a downstream dependency slightly (the [within](https://github.com/py-econometrics/within) defaults), which impacts the API changes / requires me to partially revert it.
